src/main/java/spoon/support/compiler/jdt/JDTTreeBuilderHelper.java

```diff
@@ -657,4 +672,147 @@ public boolean onAccess(char[][] tokens, int index) {
 
 		return type;
 	}
+
+	/**
+	 * Creates an entire object CtModule from a module declaration.
+	 * @return a CtModule
+	 */
+	CtModule createModule(ModuleDeclaration moduleDeclaration) {
+		CtModule module = jdtTreeBuilder.getFactory().Module().getOrCreate(new String(moduleDeclaration.moduleName));
+		module.setIsOpenModule(moduleDeclaration.isOpen());
+		jdtTreeBuilder.getContextBuilder().enter(module, moduleDeclaration);
+
+		if (moduleDeclaration.requires != null && moduleDeclaration.requires.length > 0) {
+			List<CtModuleRequirement> moduleRequirements = new ArrayList<>();
+			for (RequiresStatement requiresStatement : moduleDeclaration.requires) {
+				moduleRequirements.add(this.createModuleRequirement(requiresStatement));
+			}
+
+			module.setRequiredModules(moduleRequirements);
+		}
+
+		if (moduleDeclaration.exports != null && moduleDeclaration.exports.length > 0) {
+			List<CtPackageExport> moduleExports = new ArrayList<>();
+			for (ExportsStatement exportsStatement : moduleDeclaration.exports) {
+				moduleExports.add(this.createModuleExport(exportsStatement));
+			}
+
+			module.setExportedPackages(moduleExports);
+		}
+
+		if (moduleDeclaration.opens != null && moduleDeclaration.opens.length > 0) {
+			List<CtPackageExport> moduleOpens = new ArrayList<>();
+			for (OpensStatement opensStatement : moduleDeclaration.opens) {
+				moduleOpens.add(this.createModuleExport(opensStatement));
+			}
+
+			module.setOpenedPackages(moduleOpens);
+		}
+
+		if (moduleDeclaration.uses != null && moduleDeclaration.uses.length > 0) {
+			List<CtUsedService> consumedServices = new ArrayList<>();
+			for (UsesStatement consumedService : moduleDeclaration.uses) {
+				consumedServices.add(this.createUsedService(consumedService));
+			}
+
+			module.setUsedServices(consumedServices);
+		}
+
+		if (moduleDeclaration.services != null && moduleDeclaration.services.length > 0) {
+			List<CtProvidedService> moduleProvidedServices = new ArrayList<>();
+			for (ProvidesStatement providesStatement : moduleDeclaration.services) {
+				moduleProvidedServices.add(this.createModuleProvidedService(providesStatement));
+			}
+
+			module.setProvidedServices(moduleProvidedServices);
+		}
```

Comment:
The order of ModuleDirectives in source code is not kept. May be we might use ModuleStatement#declarationSourceStart to sort them after they are added into CtModule.

Reply:
I will create a unit test to check that and fix it

Reply:
This is managed directly in CtModuleImpl: I'm using a sorted list which uses the position to put elements in the right order. This is checked by a test.